benchmark/src/main/java/com/linecorp/decaton/benchmark/DecatonRunner.java

```diff
@@ -113,22 +125,69 @@ public String get(String key) {
                 .stateListener(state -> {
                     if (state == SubscriptionStateListener.State.RUNNING) {
                         startLatch.countDown();
                     }
                 })
                 .build();
-        resourceTracker.track(subscription.getId());
         subscription.start();
 
         startLatch.await();
     }
 
+    /**
+     * This value comes from VirtualThread#createDefaultScheduler(), which defaults keep-alive of ForkJoinPool
+     * to 30 seconds (as of Java21).
+     */
+    private static final long DRIP_INTERVAL = 30;
+    /**
+     * The intention of this method is to setup a scheduled work submitting some VirtualThread instances
+     * periodically, in order to make sure the {@link ForkJoinPool} to keep-alive the created carrier thread
+     * during warmup phase.
+     * It is necessary because {@link #onWarmupComplete(ResourceTracker)} depends on the alive list of
+     * ForkJoinPool threads to observe cpu time and memory allocation profile, which cannot be obtained from
+     * VirtualThread instances directly.
+     */
+    private void maybeSetupForkJoinPoolDrip() {
+        if (subPartitionRuntime != SubPartitionRuntime.VIRTUAL_THREAD) {
+            return;
+        }
+        ScheduledExecutorService executor = Executors.newScheduledThreadPool(
+                1, Thread.ofPlatform().daemon().factory());
+        executor.scheduleAtFixedRate(() -> {
+            for (int i = 0; i < Runtime.getRuntime().availableProcessors(); i++) {
+                Thread.startVirtualThread(() -> {
+                    try {
+                        TimeUnit.SECONDS.sleep(DRIP_INTERVAL);
+                    } catch (InterruptedException e) {
+                        throw new RuntimeException(e);
+                    }
+                });
+            }
+        }, 0, DRIP_INTERVAL, TimeUnit.SECONDS);
+    }
+
+    @Override
+    public void onWarmupComplete(ResourceTracker resourceTracker) {
+        // As we support VIRTUAL_THREAD runtime now, we can't use the technique to obtain the target Thread ID
+        // in the execution context, as in VirtualThread gets created and discarded for all tasks separately.
+        // Instead we need to observe resource usage based on the carrier thread (platform thread) side, while
+        // stdlib doesn't provide a way to resolve the carrier thread's ID.
+        ThreadMXBean threadMxBean = ResourceTracker.getSunThreadMxBean();
+        ThreadInfo[] threadInfos = threadMxBean.dumpAllThreads(true, true);
+        for (ThreadInfo threadInfo : threadInfos) {
+            String name = threadInfo.getThreadName();
+            log.debug("Tracking target check for thread name: {}", name);
+            if (name.startsWith("DecatonSubscriptionThread-") || name.startsWith("PartitionProcessorThread-") ||
+                subPartitionRuntime == SubPartitionRuntime.VIRTUAL_THREAD && name.startsWith("ForkJoinPool-")) {
```

Comment:
So if we have another fork join pool (though currently we don't) in benchmark runtime or somewhere, it may be miscounted as the resource usage.

I think it's acceptable compromise for now though, can we identify virtual thread runtime's pool more precisely somehow? (e.g. by customizing carrier thread's thread factory?)

Reply:
None without any sorts of dirty work :(

Reply:
I see, thanks. The only way I can imagine (substituting VirtualThread's scheduler by accessing package-private class) is also dirty so let's do in current way for now